packages/next/src/client/components/react-dev-overlay/ui/components/errors/dev-tools-indicator/dev-tools-info/segment-explorer.tsx

```diff
@@ -0,0 +1,199 @@
+import type { HTMLProps } from 'react'
+import { css } from '../../../../../utils/css'
+import type { DevToolsInfoPropsCore } from './dev-tools-info'
+import { DevToolsInfo } from './dev-tools-info'
+import { cx } from '../../../../utils/cx'
+import { LeftArrow } from '../../../../icons/left-arrow'
+import { useSegmentTreeClientState } from '../../../../../../../../shared/lib/devtool/app-segment-tree'
+import type {
+  Trie,
+  TrieNode,
+} from '../../../../../../../../shared/lib/devtool/trie'
+
+const IconLayout = (props: React.SVGProps<SVGSVGElement>) => {
+  return (
+    <svg
+      {...props}
+      viewBox="0 0 16 16"
+      fill="none"
+      xmlns="http://www.w3.org/2000/svg"
+    >
+      <path
+        d="M16 12.5L15.9873 12.7559C15.8677 13.9323 14.9323 14.8677 13.7559 14.9873L13.5 15H2.5L2.24414 14.9873C1.06772 14.8677 0.132274 13.9323 0.0126953 12.7559L0 12.5V1H16V12.5ZM1.5 6.25488V12.5C1.5 13.0523 1.94772 13.5 2.5 13.5H4.99512V6.25488H1.5ZM6.24512 6.25488V13.5H13.5C14.0523 13.5 14.5 13.0523 14.5 12.5V6.25488H6.24512ZM1.5 5.00488H14.5V2.5H1.5V5.00488Z"
+        fill="#666666"
+      />
+    </svg>
+  )
+}
+
+const IconPage = (props: React.SVGProps<SVGSVGElement>) => {
+  return (
+    <svg
+      {...props}
+      viewBox="0 0 16 16"
+      fill="none"
+      strokeLinejoin="round"
+      xmlns="http://www.w3.org/2000/svg"
+    >
+      <path
+        fillRule="evenodd"
+        clipRule="evenodd"
+        d="M14.5 6.5V13.5C14.5 14.8807 13.3807 16 12 16H4C2.61929 16 1.5 14.8807 1.5 13.5V1.5V0H3H8H9.08579C9.351 0 9.60536 0.105357 9.79289 0.292893L14.2071 4.70711C14.3946 4.89464 14.5 5.149 14.5 5.41421V6.5ZM13 6.5V13.5C13 14.0523 12.5523 14.5 12 14.5H4C3.44772 14.5 3 14.0523 3 13.5V1.5H8V5V6.5H9.5H13ZM9.5 2.12132V5H12.3787L9.5 2.12132Z"
+        fill="currentColor"
+      />
+    </svg>
+  )
+}
+
+const ICONS = {
+  layout: <IconLayout width={16} />,
+  page: <IconPage width={16} />,
+}
+
+function PageSegmentTree({ tree }: { tree: Trie | undefined }) {
+  if (!tree) {
+    return null
+  }
+  return (
+    <div className="segment-explorer-content">
+      <PageSegmentTreeLayerPresentation
+        tree={tree}
+        node={tree.getRoot()}
+        level={0}
+      />
+    </div>
+  )
+}
+
+function PageSegmentTreeLayerPresentation({
+  tree,
+  node,
+  level,
+}: {
+  tree: Trie
+  node: TrieNode
+  level: number
+}) {
+  const segments = node.value?.split('/') || []
+  const fileName = segments[segments.length - 1] || ''
+  const nodeName = fileName.split('.')[0] === 'layout' ? 'layout' : 'page'
```

Comment:
since we construct the tree and we know in `create-component-tree` whether the segment is a layout or page, could we explicitly pass this rather than deriving it from the filename? Filename parsing feels brittle